### PR TITLE
chore(metrics): deprecate withMetricLogger in favor of withMetricsLogger

### LINKED ISSUE #1059: Rename MetricsUtils.withMetricLogger to MetricsUtils.withMetricsLogger
https://github.com/awslabs/aws-lambda-powertools-java/pull/1000 introduced the wrongly named `MetricsUtils.withMetricLogger`. To be consistent, it should have been called `MetricsUtils.withMetricsLogger`.

Deprecate `MetricsUtils.withMetricLogger` and introduce `MetricsUtils.withMetricsLogger`.

### PR DESCRIPTION
**Pull request by humanzz**
**Issue #:** #1059

## Description of changes:

- deprecate `MetricsUtils.withMetricLogger`
- introduce `MetricsUtils.withMetricsLogger`

**Checklist**

 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-java/#tenets)
* [x] Update tests
* [ ] Update docs
* [x] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/)

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsUtils.java [hunk 1]
```diff
@@ -130,6 +130,20 @@ public static void withMetricLogger(final Consumer<MetricsLogger> logger) {
...
+     * specified either on {@link Metrics} annotation or via POWERTOOLS_METRICS_NAMESPACE env var.
+     * It by default captures function_request_id as property if used together with {@link Metrics} annotation. It will also
+     * capture xray_trace_id as property if tracing is enabled.
+     *
+     * @param logger the MetricsLogger
+     *
+     * @deprecated use {@link MetricsUtils#withMetricsLogger} instead
+     */
+    public static void withMetricLogger(final Consumer<MetricsLogger> logger) {
+        withMetricLogger(logger);
```
Comment (jeromevdl) [inline powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsUtils.java:144]: infinite loop

```suggestion
        withMetricsLogger(logger);
```

**Review by jeromevdl — CHANGES_REQUESTED** [review]
Please fix the infinite loop.

File: powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsUtils.java (hunk 1 above)
Comment (humanzz) [reply powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsUtils.java:144]: oops!

**Review by jeromevdl — DISMISSED** [review]
OK

File: powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsUtils.java [hunk 2]
```diff
@@ -130,6 +130,20 @@ public static void withMetricLogger(final Consumer<MetricsLogger> logger) {
...
+     * Provide and immediately flush a {@link MetricsLogger}. It uses the default namespace
+     * specified either on {@link Metrics} annotation or via POWERTOOLS_METRICS_NAMESPACE env var.
+     * It by default captures function_request_id as property if used together with {@link Metrics} annotation. It will also
+     * capture xray_trace_id as property if tracing is enabled.
+     *
+     * @param logger the MetricsLogger
+     *
+     * @deprecated use {@link MetricsUtils#withMetricsLogger} instead
+     */
+    public static void withMetricLogger(final Consumer<MetricsLogger> logger) {
```
Comment (kozub) [inline powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsUtils.java:143]: Does it make sense to add also `@Deprecated` annotation?

Comment (jeromevdl) [reply powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsUtils.java:143]: Sure!

Comment (jeromevdl) [inline powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsUtils.java:143]:
```suggestion
    @Deprecated
    public static void withMetricLogger(final Consumer<MetricsLogger> logger) {
```

**Review by kozub — APPROVED** [review]

File: powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/MetricsLoggerTest.java [hunk 3]
```diff
@@ -124,13 +126,29 @@ void singleMetricsCaptureUtilityWithDefaultNameSpace() {
...
+    }
+
+    @Test
+    void shouldThrowExceptionWhenDefaultDimensionIsNull() {
+        assertThatNullPointerException()
+                .isThrownBy(() -> MetricsUtils.defaultDimensionSet(null))
+                .withMessage("Null dimension set not allowed");
+    }
+
+    private void testWithNewLogger(Consumer<Consumer<MetricsLogger>> methodToTest) {
```
Comment (jeromevdl) [inline powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/MetricsLoggerTest.java:144]: strange to name it `NewLogger` when you use it with the deprecated one...

```suggestion
    private void testLogger(Consumer<Consumer<MetricsLogger>> loggerToTest) {
```

File: powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/MetricsLoggerTest.java [hunk 4]
```diff
@@ -154,13 +172,6 @@ void metricsLoggerCaptureUtilityWithDefaultNameSpace() {
         }
     }
 
-    @Test
```
Comment (jeromevdl) [inline powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/MetricsLoggerTest.java:157]: why did you remove this test ?

File: powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/MetricsLoggerTest.java (hunk 3 above)
Comment (humanzz) [reply powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/MetricsLoggerTest.java:144]: New was as opposed to the default `MetricsUtils.metricsLogger()` and `MetricsUtils.withMetric(s)Logger` essentially providing new ones. However, happy to call it `testLogger` instead to avoid any confusion

File: powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/MetricsLoggerTest.java (hunk 4 above)
Comment (humanzz) [reply powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/MetricsLoggerTest.java:157]: Oh wow, I'm on a roll with mistakes here. Unintentional as I was moving methods around

Comment (humanzz) [reply powertools-metrics/src/test/java/software/amazon/lambda/powertools/metrics/MetricsLoggerTest.java:157]: It's not removed, check line 138

**Comment by humanzz** [comment]
Any further things I might have missed?

**Review by jeromevdl — APPROVED** [review]
